File: code/646.maximum-length-of-pair-chain.0.py
class Solution:
    def findLongestChain(self, pairs: List[List[int]]) -> int:
        pairs.sort(key = lambda o : (o[0], o[1]))
        p = []
        for pair in pairs:
            if len(p) == 0:
                p.append(pair[1])
            else:
                idx = bisect.bisect_left(p, pair[0])
                if idx == len(p):
                    p.append(pair[1])
                else:
                    p[idx] = min(p[idx], pair[1])
        return len(p)
            
        

File: code/646.maximum-length-of-pair-chain.0.cpp
class Solution {
 public:
  int findLongestChain(vector<vector<int>>& pairs) {
    struct {
      mutable std::optional<int> f[1005] = {};
    } mempool;

    const int n = std::size(pairs);

    const auto sorted_pairs = [&](std::vector<std::vector<int>> self = {}) {
      std::copy(std::begin(pairs), std::end(pairs), std::back_inserter(self));
      std::sort(std::begin(self), std::end(self));
      return self;
    }();    

    std::function<int(int)> f = [&, memo = mempool.f](int i) {
      if (memo[i].has_value()) {
        return memo[i].value();
      }
      else {
        return memo[i].emplace([&] {
          if (i == 0) {
            return 1;
          }
          else {
            return [&](int acc = 1) {
              for (int j = 0; j < i; ++j) {            
                if (sorted_pairs[j][1] < sorted_pairs[i][0]) {
                  acc = std::max(acc, 1 + f(j));
                }
              }
              return acc;
            }();
          }
        }());
      }
    };

    auto solve = [&](int acc = 0) {
      for (int i = 0; i < n; ++i) {
        acc = std::max(acc, f(i));
      }
      return acc;
    };

    return solve();
  }
};
